### PR TITLE
ZEPPELIN-143: Git as a versioned notebook storage

### PR DESCRIPTION
**Pull request by bzz**
This is very basic implementation of the [ZEPPELIN-143](https://issues.apache.org/jira/browse/ZEPPELIN-143) at the backend.

It makes a local git repository our of your `/notebook` dir and commits a new revision for each save/update.

It does not: 
 - add any remotes to the git repo. It is totally possible to do that manually though. It would be interesting to add this later, to be able to push the notebook to hostings like GH
 - have any GUI modifications. It is left as further work, to add the ability for a user to switch "versions" of the notebook,  navigating between previous runs.

Feedback is very welcome!

### CODE REVIEW

**Comment by bzz** [comment]
\cc @Leemoonsoo @jongyoul @felixcheung @khalidhuseynov for a review

**Comment by bzz** [comment]
CI fails with

```
Results :

Failed tests: 
  NotebookTest.testAbortParagraphStatusOnInterpreterRestart:347 expected:<ABORT> but was:<RUNNING>

Tests run: 34, Failures: 1, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Zeppelin ........................................... SUCCESS [ 12.289 s]
[INFO] Zeppelin: Interpreter .............................. SUCCESS [01:18 min]
[INFO] Zeppelin: Zengine .................................. FAILURE [  8.411 s]

.....

Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.662 sec <<< FAILURE! - in org.apache.zeppelin.notebook.NotebookTest
testAbortParagraphStatusOnInterpreterRestart(org.apache.zeppelin.notebook.NotebookTest)  Time elapsed: 0.108 sec  <<< FAILURE!
java.lang.AssertionError: expected:<ABORT> but was:<RUNNING>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.apache.zeppelin.notebook.NotebookTest.testAbortParagraphStatusOnInterpreterRestart(NotebookTest.java:347)
```

**Comment by khalidhuseynov** [comment]
This CI problem should be solved by #468, may need to rebase for that

File: docs/storage/storage.md [hunk 1]
```diff
@@ -19,10 +19,27 @@ limitations under the License.
 -->
 ### Notebook Storage
 
-In Zeppelin there are two option for storage Notebook, by default the notebook is storage in the notebook folder in your local File System and the second option is S3.
+Zeppelin a pluggable notebook storage mechanism with multiple implementations.
+There are few Notebook storage options avaialble for a use,
```
Comment (felixcheung) [inline docs/storage/storage.md]: Is it possible to have multiple storage simultaneously? Could we clarify in this doc?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java [hunk 2]
```diff
@@ -282,7 +294,8 @@ private void printParagraph(Paragraph paragraph) {
     LOG.info("Paragraph ID : " + paragraph.getId());
     LOG.info("Paragraph title : " + paragraph.getTitle());
   }
-  
+
+  @SuppressWarnings("unused")
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java]: why was there a uncalled private method?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+
+/**
+ * NotebookRepo that hosts all the notebook FS in a single Git repo
+ *
+ * This impl intended to be simple and straightforward:
+ *   - does not handle branches
+ *   - only basic local git file repo, no remote Github push\pull yet
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java]: would we be handling push in the future? if so, how would it fit into the API or would there be new API for checkpointing/pushing to remote?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  public GitNotebookRepo(ZeppelinConfiguration conf) throws IOException {
+    super(conf);
+    localPath = getRootDir().getName().getPath();
+    LOG.info("Opening a git repo at '{}'", localPath);
+    Repository localRepo = new FileRepository(Joiner.on(File.separator).join(localPath, ".git"));
+    if (!localRepo.getDirectory().exists()) {
+      LOG.info("Git repo {} does not exist, creating a new one", localRepo.getDirectory());
+      localRepo.create();
+    }
+    git = new Git(localRepo);
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java]: does the localRepo need to be `close`d to release resource? http://download.eclipse.org/jgit/docs/jgit-2.0.0.201206130900-r/apidocs/org/eclipse/jgit/lib/Repository.html#close()

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  private void maybeAddAndCommit(String pattern) {
+    try {
+      List<DiffEntry> gitDiff = git.diff().call();
+      if (!gitDiff.isEmpty()) {
+        LOG.debug("Changes found for pattern '{}': {}", pattern, gitDiff);
+        DirCache added = git.add().addFilepattern(pattern).call();
+        LOG.debug("{} changes are about to be commited", added.getEntryCount());
+        git.commit().setMessage("Updated " + pattern).call();
+      } else {
+        LOG.info("No changes found {}", pattern);
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java]: probably should be LOG.verbose or debug...

**Comment by felixcheung** [comment]
looks good! thanks for working on this. looking forward to git push ;)

**Comment by jeffsteinmetz** [comment]
Is there a document (or could one be added) that describes setting up the Git repo integration (git ssh or https path, connecting Zeppelin to a specific git repo, passwords, ssh keys, etc.)?

**Comment by bzz** [comment]
Guys, thanks for reviews, appreciate that a lot!

Will address each of them in next couple of days.

**Comment by felixcheung** [comment]
@jeffsteinmetz I think this is committing to local only, no remote repo support yet

**Comment by bzz** [comment]
@jeffsteinmetz @felixcheung well noted, this PR introduces automation to keep notebook dir inside the git repository in your local filesystem.

Right now, this repository does not have any 'remotes' setup, and Zeppelin does not push it anywhere, so there are no Zeppelin configuration needed (except documented property change).

It is possible and quite simple to sync this local repo manually by `cd notebooks; git add remote ...; git push`, as well as I can add this ability to Zeppelin itself, if people find it useful (in the separate PR).

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java (hunk 3 above)
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java]: Checkpointing happens in Git on commit level, so I was thinking that push, if have enough user demand, can be implemented on the NotebookRepoSync level, like a FS->S3 sync does.
This should not require big API changes here.

**Comment by bzz** [comment]
@khalidhuseynov thanks for looking into CI problem, that's a bummer!
> This CI problem should be solved by #468, may need to rebase for that

It looks like it was merged 6 days ago, so current branch should be on top of it already, but just in case - I have synced with the latest master and it seems to pass! If that repeats - will file the jira issue with label 'flaky-test'.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java (hunk 2 above)
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java]: My guess that those procedures were left by the original author as a debug facilities that were useful at some time during the development

File: docs/storage/storage.md (hunk 1 above)
Comment (felixcheung) [inline docs/storage/storage.md]: ... multiple storage *enabled* simultaneously...

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java (hunk 3 above)
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java]: +1 for remote push :)

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java (hunk 4 above)
Comment (bzz) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java]: Very good point, thank you! addressed in 5d7ffea

**Comment by bzz** [comment]
In 5d7ffea  `.close()` lifecycle method was introduced for the `NotebookRero` to cleanup the resources that storage implementation might have occupied.
As this is an API change \cc @khalidhuseynov and @Leemoonsoo who worked on this before, as I recall, to take a quick look.

**Comment by bzz** [comment]
New CI failure - cassandra interpreter tests

```
[INFO] Zeppelin: Lens interpreter ......................... SUCCESS [  3.570 s]
[INFO] Zeppelin: Cassandra ................................ FAILURE [ 33.880 s]

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 21.414 sec <<< FAILURE! - in org.apache.zeppelin.cassandra.CassandraInterpreterTest
org.apache.zeppelin.cassandra.CassandraInterpreterTest  Time elapsed: 21.414 sec  <<< ERROR!
java.lang.ExceptionInInitializerError: null
	at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:84)
	at com.datastax.driver.core.DefaultResultSetFuture.extractCauseFromExecutionException(DefaultResultSetFuture.java:269)
	at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:183)
	at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:52)
	at info.archinnov.achilles.script.ScriptExecutor.executeScript(ScriptExecutor.java:58)
	at info.archinnov.achilles.embedded.AchillesInitializer.executeStartupScripts(AchillesInitializer.java:194)
	at info.archinnov.achilles.embedded.AchillesInitializer.initialize(AchillesInitializer.java:97)
	at info.archinnov.achilles.embedded.AchillesInitializer.initializeFromParameters(AchillesInitializer.java:60)
	at info.archinnov.achilles.embedded.CassandraEmbeddedServer.<init>(CassandraEmbeddedServer.java:76)
	at info.archinnov.achilles.junit.AchillesResource.initResource(AchillesResource.java:59)
	at info.archinnov.achilles.junit.AchillesResource.<init>(AchillesResource.java:52)
	at info.archinnov.achilles.junit.AchillesResourceBuilder.build(AchillesResourceBuilder.java:205)
	at org.apache.zeppelin.cassandra.CassandraInterpreterTest.<clinit>(CassandraInterpreterTest.java:62)
Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: localhost/127.0.0.1:9267 (com.datastax.driver.core.OperationTimedOutException: [localhost/127.0.0.1:9267] Operation timed out))
	at com.datastax.driver.core.RequestHandler.reportNoMoreHosts(RequestHandler.java:216)
	at com.datastax.driver.core.RequestHandler.access$900(RequestHandler.java:45)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.sendRequest(RequestHandler.java:276)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution$1.run(RequestHandler.java:374)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by bzz** [comment]
re-triggering CI build

**Comment by bzz** [comment]
CI passes now, ready to be merged.

**Comment by khalidhuseynov** [comment]
@bzz thanks for starting this feature! `.close()` is required in this case and is fine to add to API. LGTM

**Comment by Leemoonsoo** [comment]
@bzz Tested and working really well. And the API looks fine.

Could you also take care `zeppelin-distribution/src/bin_license/LICENSE` for newly added dependency, jgit? jgit and it's transitive dependency looks like

```
+- org.eclipse.jgit:org.eclipse.jgit:jar:4.1.1.201511131810-r:compile
  +- com.jcraft:jsch:jar:0.1.53:compile
  +- com.googlecode.javaewah:JavaEWAH:jar:0.7.9:compile
  +- org.apache.httpcomponents:httpclient:jar:4.3.6:compile
  \- org.eclipse.jdt:org.eclipse.jdt.annotation:jar:1.1.0:compile
```

**Comment by bzz** [comment]
@Leemoonsoo thank you for kind reminder! Will do it and ping back

**Comment by bzz** [comment]
Thank all participants for kind reviews!

@Leemoonsoo could you check 468a858 and let me know if you think that is enough?

**Comment by Leemoonsoo** [comment]
Thanks @bzz Looks good to me!

**Comment by bzz** [comment]
Thank you!
Merging if there is no more discussion

**Comment by p0wl** [comment]
Thanks for the feature!

Any news on automatic github integration? Or a guidline how to setup github sync?

**Comment by herval** [comment]
@p0wl I started remote git support here: https://github.com/herval/zeppelin/tree/remote-git-support - it's a WIP, if someone wants to take over (it would work w/ github or any other git repo)

**Comment by Gauravshah** [comment]
@herval were you able to finish it ?

**Comment by herval** [comment]
Haven't had the time to look into that further. TBH, I don't think storing notebooks as text files is a good strategy - not in the environment we use Zeppelin, anyway. We're likely moving towards some sort of centralized storage (maybe a KV storage), in order to be able to scale horizontally

**Pull request closed without merging**